app/controllers/concerns/saml_idp_logout_concern.rb

```diff
@@ -1,129 +1,38 @@
-# rubocop:disable ModuleLength
 # Although this module is long all the code is related to the SLO concern.
-require 'saml_idp/logout_request_builder'
 require 'saml_idp/logout_response_builder'
 
 module SamlIdpLogoutConcern
   extend ActiveSupport::Concern
 
-  included do
-    before_action :validate_saml_logout_param, only: :logout
-  end
-
-  def saml_logout_message
-    return handle_saml_logout_response if successful_saml_response?
-    return nil if finish_logout_at_idp?
-    return handle_saml_logout_request(name_id_user) if valid_saml_request?
-
-    generate_slo_request(current_user)
-  end
-
   private
 
-  def finish_logout_at_idp?
-    !user_signed_in_and_has_identity? || failed_saml_response? || slo_not_implemented_at_sp?
-  end
-
-  def slo_not_implemented_at_sp?
-    slo_identity = fetch_identity_for_slo(current_user)
-    slo_identity.sp_metadata[:assertion_consumer_logout_service_url].nil?
-  end
-
-  def successful_saml_response?
-    @saml_response.present? && @saml_response.success?
-  end
-
-  def failed_saml_response?
-    @saml_response.present? && !@saml_response.success?
-  end
-
-  def valid_saml_request?
-    saml_request.present? && saml_request.valid?
-  end
-
-  def user_signed_in_and_has_identity?
-    user_signed_in? && current_user.active_identities.present?
+  def slo
+    @_slo ||= SingleLogoutHandler.new(@saml_response, saml_request, current_user)
   end
 
   def handle_saml_logout_response
-    unless asserted_identity.nil?
-      resource = asserted_identity.user
-      asserted_identity.deactivate!
-    end
-
-    # continue logout with next identity if present
-    return generate_slo_request(resource) if resource_in_slo?(resource)
-
-    # no SLO messages to generate; finish logout at IdP
-    return nil if user_session[:logout_response].nil?
+    LogoutResponseHandler.new(asserted_identity, user_session[:logout_response]).perform do |status|
+      return generate_slo_request if status == 'continue logout with next identity'
```

Comment:
The `LogoutResponseHandler` is yielding the state of the SLO flow to the controller so that the controller knows how to proceed. The status can be anything, and the controller needs to check the value of the status. The easiest thing to compare is either a string or a symbol. I thought a string was the most descriptive, but we could also do something like `if status == :continue_logout_with_next_identity`